x-pack/plugins/security_solution/server/lib/entity_analytics/asset_criticality/asset_criticality_service.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEmpty } from 'lodash/fp';
+import type { AssetCriticalityRecord } from '../../../../common/api/entity_analytics';
+import type { AssetCriticalityDataClient } from './asset_criticality_data_client';
+
+interface CriticalityIdentifier {
+  id_field: string;
+  id_value: string;
+}
+
+interface IdentifierValuesByField {
+  [idField: string]: string[];
+}
+
+export interface AssetCriticalityService {
+  getCriticalitiesByIdentifiers: (
+    identifiers: CriticalityIdentifier[]
+  ) => Promise<AssetCriticalityRecord[]>;
+}
+
+const isCriticalityIdentifierValid = (identifier: CriticalityIdentifier): boolean =>
+  !isEmpty(identifier.id_field) && !isEmpty(identifier.id_value);
+
+const groupIdentifierValuesByField = (
+  identifiers: CriticalityIdentifier[]
+): IdentifierValuesByField =>
+  identifiers.reduce((acc, id) => {
+    if (!acc[id.id_field]) {
+      acc[id.id_field] = [];
+    }
+    if (!acc[id.id_field].includes(id.id_value)) {
+      acc[id.id_field].push(id.id_value);
+    }
+    return acc;
+  }, {} as IdentifierValuesByField);
+
+const buildCriticalitiesQuery = (identifierValuesByField: IdentifierValuesByField) => ({
```

Comment:
Instead of building the query here and supplying to the generic `assetCriticalityDataClient.search` method, would it be neater for the client to have responsibility for building the query and expose a method like `multiGet` which just takes the ID fields and values? 

In my mind the client is the one that "knows" about the asset criticality index.

We could also then use the `_mget` API to just grab the docs by ID, it would have the benefit of not being affected by refresh intervals (e.g serverless). But I don't know how you feel about tying ourselves to the "known ID" implementation as we know it only works with  a plain index and therefore may change in the future.

Reply:
That's interesting; I was thinking about it like this:

* DataClient is dumb, and does what you tell it
    * It knows _where_ data is, how/when to bootstrap data, and (future) how/when to upgrade or do other lifecycle stuff
* Service knows how to construct a query based on identifiers
* Caller of Service just knows that it wants records based on identifiers

The purpose (to me) of the `AssetCriticalityService` is to provide a declarative, straightforward interface for common criticality tasks. If we moved the query generation to the DataClient, the Service really wouldn't be doing anything.

> We could also then use the _mget API to just grab the docs by ID

It sounds like we're in agreement about _how_ this should work (`fnA = (ids) => fnB(buildQuery(ids))`); you're just proposing that `fnA` and `fnB` live in the `DataClient` while I've got `fnA` in a separate service.

>  it would have the benefit of not being affected by refresh intervals (e.g serverless)

P.S. we ran into this during integration testing and it took me a while to figure out what was going on 😅 ; see `waitForAssetCriticalityToBePresent`